### PR TITLE
Cli.UnitTests: template resolution tests refactoring

### PR DESCRIPTION
**Pull request by vlada-shubina**
- refactored `MockTemplateInfo` and `MockNewCommandInput` mock classes 
  - made them ready to be used in data driven tests (`IXUnitSerializable` interface implementation)
  - made construction easier (allowed to reduce the code by 1k+ lines)

Some tests will be converted to data driven tests in next PR

### CODE REVIEW

**Review by DavidKarlas — APPROVED** [review]

File: test/Microsoft.TemplateEngine.Cli.UnitTests/CliMocks/MockNewCommandInput.cs [hunk 1]
```diff
@@ -5,86 +5,173 @@
...
+        public bool IsListFlagSpecified => _commandOptions.ContainsKey("--list") || _commandOptions.ContainsKey("-l");
 
-        public bool IsQuietFlagSpecified { get; set; }
+        public bool IsQuietFlagSpecified { get; }
 
-        public bool IsShowAllFlagSpecified { get; set; }
+        public bool IsShowAllFlagSpecified { get; }
 
-        public string Language { get; set; }
+        public string Language
```
Comment (donJoseLuis) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/CliMocks/MockNewCommandInput.cs:146]: Most properties in this Type boast the same pattern & could all be simplified via ternary operator
`public string Language => _commandOptions.ContainsKey("--language") ? _commandOptions["--language"] : string.Empty;`

File: test/Microsoft.TemplateEngine.Cli.UnitTests/CliMocks/MockNewCommandInput.cs [hunk 2]
```diff
@@ -209,5 +306,28 @@ public IReadOnlyList<string> VariantsForCanonical(string canonical)
...
         }
+
+        #region IXunitSerializable implementation
+        public void Deserialize(IXunitSerializationInfo info)
+        {
+            TemplateName = info.GetValue<string>("command_templateName");
+            _commandOptions = JsonConvert.DeserializeObject<Dictionary<string, string>>(info.GetValue<string>("command_options"));
+            _templateOptions = JsonConvert.DeserializeObject<Dictionary<string, string>>(info.GetValue<string>("command_templateOptions"));
+        }
+        public void Serialize(IXunitSerializationInfo info)
```
Comment (donJoseLuis) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/CliMocks/MockNewCommandInput.cs:317]: minor: 1 line between methods

File: test/Microsoft.TemplateEngine.Cli.UnitTests/CliMocks/MockNewCommandInput.cs [hunk 3]
```diff
@@ -209,5 +306,28 @@ public IReadOnlyList<string> VariantsForCanonical(string canonical)
...
+        {
+            info.AddValue("command_templateName", TemplateName, typeof(string));
+            info.AddValue("command_options", JsonConvert.SerializeObject(_commandOptions), typeof(string));
+            info.AddValue("command_templateOptions", JsonConvert.SerializeObject(_templateOptions), typeof(string));
+        }
+
+        public override string ToString()
```
Comment (donJoseLuis) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/CliMocks/MockNewCommandInput.cs:324]: minor: unless this string output is very necessary, consider simplifying it, specially if called often.  Why?  strings :(

File: test/Microsoft.TemplateEngine.Cli.UnitTests/TemplateResolutionTests/TemplateMatchInfoExtensionTests.cs [hunk 4]
```diff
@@ -32,12 +28,7 @@ public void TestHasClassificationMatchAndNameMismatch_ShortNameMismatchAndClassi
                 Kind = MatchKind.Exact
             });
 
-            ITemplateMatchInfo testTemplate = new TemplateMatchInfo(new TemplateInfo()
-            {
-                Name = "Template1",
-                Identity = "Template1",
-                GroupIdentity = "TestGroup"
-            }, testDisposiions);
+            ITemplateMatchInfo testTemplate = new TemplateMatchInfo(new MockTemplateInfo("Template1", groupIdentity: "TestGroup"), testDisposiions);
```
Comment (donJoseLuis) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/TemplateResolutionTests/TemplateMatchInfoExtensionTests.cs:31]: minor: small typo in `testDisposiions`; probably meant *testDispositions*

File: test/Microsoft.TemplateEngine.Cli.UnitTests/TemplateResolutionTests/TemplateMatchInfoTests.cs [hunk 5]
```diff
@@ -11,7 +12,7 @@ public class TemplateMatchInfoTests
         [Fact(DisplayName = nameof(EmptyMatchDisposition_ReportsCorrectly))]
         public void EmptyMatchDisposition_ReportsCorrectly()
         {
-            ITemplateInfo templateInfo = new TemplateInfo();
+            ITemplateInfo templateInfo = new MockTemplateInfo();
```
Comment (donJoseLuis) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/TemplateResolutionTests/TemplateMatchInfoTests.cs:15]: minor:
design question, in these cases, instantiating via new vs. factory.  If you wanted to later used distinct mocks, you'd not have to refactor.

File: test/Microsoft.TemplateEngine.Cli.UnitTests/TemplateSearchCacheTests.cs [hunk 6]
```diff
@@ -51,15 +48,7 @@ public async Task CacheSearchCliSymbolNameFilterTest()
...
-            // The template symbol is "Framework" (capital "F"). This checks that the host specific override is applied
-            Dictionary<string, string> rawCommandInputs = new Dictionary<string, string>()
-            {
-                { "framework", "netcoreapp2.0" }
-            };
-            INewCommandInput commandInput = new MockNewCommandInput(rawCommandInputs)
-            {
-                TemplateName = "foo"
-            };
+            INewCommandInput commandInput = new MockNewCommandInput("foo").WithTemplateOption("framework", "netcoreapp2.0");
```
Comment (donJoseLuis) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/TemplateSearchCacheTests.cs:51]: minor:
should these values come from a test data set or config? (i.e., `framework` & `netcoreapp2.0`)

File: test/Microsoft.TemplateEngine.Cli.UnitTests/TemplateSearchCacheTests.cs [hunk 7]
```diff
@@ -260,79 +210,28 @@ public async Task CacheSearchLanguageMismatchFilterTest()
...
-            Tags = new Dictionary<string, ICacheTag>(StringComparer.Ordinal)
-            {
-                {
-                    "language", ResolutionTestHelper.CreateTestCacheTag(new List<string>() { "F#" })
-                }
-            },
-            CacheParameters = new Dictionary<string, ICacheParameter>()
-        };
+        private static readonly ITemplateInfo _barCSharpTemplate = new MockTemplateInfo("barC", name: "MockBarCsharpTemplate", identity: "Mock.Bar.1.Csharp", groupIdentity: "Mock.Bar")
+                                                                        .WithDescription("Mock Bar CSharp template")
```
Comment (donJoseLuis) [inline test/Microsoft.TemplateEngine.Cli.UnitTests/TemplateSearchCacheTests.cs:228]: minor:
consider consistent decorator formatting.  i other places the .With* decorators boast different spacing formatting with respect to the decoration target.

File: test/Microsoft.TemplateEngine.Mocks/MockTemplateInfo.cs [hunk 8]
```diff
@@ -1,87 +1,342 @@
...
+        public IReadOnlyList<ITemplateParameter> Parameters
+        {
+            get
+            {
+                if (_parameters == null)
+                {
+                    List<ITemplateParameter> parameters = new List<ITemplateParameter>();
 
-        public IReadOnlyDictionary<string, ICacheTag> Tags { get; set; }
+                    foreach (KeyValuePair<string, ICacheTag> tagInfo in Tags)
```
Comment (donJoseLuis) [inline test/Microsoft.TemplateEngine.Mocks/MockTemplateInfo.cs:172]: minor:
feels like the content of this get should be modularized

File: test/Microsoft.TemplateEngine.Mocks/MockTemplateInfo.cs [hunk 9]
```diff
@@ -1,87 +1,342 @@
...
+
+        public override string ToString()
+        {
+            StringBuilder sb = new StringBuilder();
+            sb.Append("Short name:" + string.Join(",", _shortNameList) + ";");
+            sb.Append("Identity:" + Identity + ";");
+
+            if (string.IsNullOrEmpty(GroupIdentity))
+            {
+                sb.Append("Group:<not set>;");
```
Comment (donJoseLuis) [inline test/Microsoft.TemplateEngine.Mocks/MockTemplateInfo.cs:306]: minor:
`StringBuilder::Append` returns `StringBuilder`.  Consider `_ = sb.Append("Group:<not set>;");`

**Review by donJoseLuis — APPROVED** [review]

File: test/Microsoft.TemplateEngine.Cli.UnitTests/CliMocks/MockNewCommandInput.cs (hunk 3 above)
Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Cli.UnitTests/CliMocks/MockNewCommandInput.cs:324]: This output will be shown in data-driven tests row later own. I would prefer better output for tests.

File: test/Microsoft.TemplateEngine.Cli.UnitTests/TemplateSearchCacheTests.cs (hunk 6 above)
Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Cli.UnitTests/TemplateSearchCacheTests.cs:51]: Test data is coming in the next PR. Many of these changes are done to enable data driven tests

File: test/Microsoft.TemplateEngine.Cli.UnitTests/TemplateResolutionTests/TemplateMatchInfoTests.cs (hunk 5 above)
Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Cli.UnitTests/TemplateResolutionTests/TemplateMatchInfoTests.cs:15]: So far there is a single class for template info, however point taken., thank you.

File: test/Microsoft.TemplateEngine.Cli.UnitTests/CliMocks/MockNewCommandInput.cs [hunk 10]
```diff
@@ -5,86 +5,173 @@
 using Microsoft.TemplateEngine.Abstractions;
 using Microsoft.TemplateEngine.Cli.CommandParsing;
 using Microsoft.TemplateEngine.Edge.Template;
+using Newtonsoft.Json;
+using Xunit.Abstractions;
 
 namespace Microsoft.TemplateEngine.Cli.UnitTests.CliMocks
 {
-    internal class MockNewCommandInput : INewCommandInput
+    internal class MockNewCommandInput : INewCommandInput, IXunitSerializable
```
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply test/Microsoft.TemplateEngine.Cli.UnitTests/CliMocks/MockNewCommandInput.cs:13]: so far there are no multi-threaded scenarios for using it.

**Pull request merged**